### PR TITLE
KTOR-5502 Fixed request in Darwin engine deadlocks on macOS since 2.2.2

### PR DESCRIPTION
**Pull request by rsinukov**
https://youtrack.jetbrains.com/issue/KTOR-5502

### CODE REVIEW

**Review by e5l — CHANGES_REQUESTED** [review]
Should we apply the same fixes for a legacy?

File: ktor-client/ktor-client-darwin/darwin/src/io/ktor/client/engine/darwin/DarwinClientEngine.kt [hunk 1]
```diff
@@ -18,7 +18,11 @@ import kotlin.coroutines.*
 
 @OptIn(InternalAPI::class)
 internal class DarwinClientEngine(override val config: DarwinClientEngineConfig) : HttpClientEngineBase("ktor-darwin") {
-    private val requestQueue: NSOperationQueue? = NSOperationQueue.currentQueue()
+
+    private val requestQueue: NSOperationQueue? = when (val queue = NSOperationQueue.currentQueue()) {
```
Comment (e5l) [inline ktor-client/ktor-client-darwin/darwin/src/io/ktor/client/engine/darwin/DarwinClientEngine.kt:22]: Maybe we should always use new queue by default

**Comment by e5l** [comment]
Please also check codestyle

**Review by e5l — APPROVED** [review]
lgtm

File: ktor-client/ktor-client-darwin/darwin/test/DarwinEngineTest.kt [hunk 2]
```diff
@@ -20,10 +19,36 @@ import kotlin.test.*
 
 class DarwinEngineTest {
 
-    val testCoroutineContext: CoroutineContext = Dispatchers.Default
+    @Test
+    fun testRequestInRunBlockingDispatchersDefault() = runBlocking(Dispatchers.Default) {
```
Comment (wkornewald) [inline ktor-client/ktor-client-darwin/darwin/test/DarwinEngineTest.kt:23]: Why is Dispatchers.Default needed here and why can't Ktor internally do the right thing, so it's safe to be called from any dispatcher/thread?

Comment (rsinukov) [reply ktor-client/ktor-client-darwin/darwin/test/DarwinEngineTest.kt:23]: What do you mean by why is it needed? This tests that request can be made on default dispatcher. Below are the tests for other dispatchers, and no dispatcher.

Comment (wkornewald) [reply ktor-client/ktor-client-darwin/darwin/test/DarwinEngineTest.kt:23]: Sorry, I didn't see the one without dispatcher. But in our unit tests everything just blocks. Only Default doesn't block.

Comment (rsinukov) [reply ktor-client/ktor-client-darwin/darwin/test/DarwinEngineTest.kt:23]: Which version are you using? This fixed will be released in 2.3.0. You can try [EAP version](https://ktor.io/eap/), if you want to check the fix.

Comment (wkornewald) [reply ktor-client/ktor-client-darwin/darwin/test/DarwinEngineTest.kt:23]: I first tried with a local build, but with 2.3.0-eap-631 the issue is indeed fixed and apparently my repo had an outdated remote for the `main` branch. 🤦  Sorry for the confusion. Everything works as expected.

Comment (rsinukov) [reply ktor-client/ktor-client-darwin/darwin/test/DarwinEngineTest.kt:23]: Great, thanks for confirming.

**Pull request merged**